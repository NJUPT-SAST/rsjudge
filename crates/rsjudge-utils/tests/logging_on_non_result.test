use std::fmt::Display;

use rsjudge_utils::log_if_error;

fn main() {
    #[derive(Debug)]
    struct S;
    impl Display for S {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "S")
        }
    }
    impl S {
        fn map_err<F: FnOnce(Self) -> Self>(self, f: F) -> Self {
            f(self)
        }
    }

    let _ = log_if_error!(S);
}
