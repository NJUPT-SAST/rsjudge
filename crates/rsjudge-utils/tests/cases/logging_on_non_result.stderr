error[E0308]: mismatched types
  --> tests/cases/logging_on_non_result.rs:21:27
   |
21 |     let _ = log_if_error!(S);
   |             --------------^-
   |             |             |
   |             |             expected `Result<_, _>`, found `S`
   |             arguments to this function are incorrect
   |
   = note: expected enum `Result<_, _>`
            found struct `S`
note: method defined here
  --> $RUST/core/src/result.rs
   |
   |     pub fn map_err<F, O: FnOnce(E) -> F>(self, op: O) -> Result<T, F> {
   |            ^^^^^^^
help: try wrapping the expression in a variant of `Result`
   |
21 |     let _ = log_if_error!(Ok(S));
   |                           +++ +
21 |     let _ = log_if_error!(Err(S));
   |                           ++++ +
